### PR TITLE
Prefer default argument in `ExecutionSpace::[impl_static_]fence`

### PR DESCRIPTION
**Pull request by dalg24**
Yet another attempt to harmonize/simplify the interface of execution spaces.
These changes are necessary to write generic code in #5144

* Adding missing `HPX::impl_static_fence`
* Drop leading execution space argument in `OpenMP::implemented_static_fence`

### CODE REVIEW

File: core/src/Kokkos_Threads.hpp [hunk 1]
```diff
@@ -107,11 +107,12 @@ class Threads {
   /// return asynchronously, before the functor completes.  This
   /// method does not return until all dispatched functors on this
   /// device have completed.
-  static void impl_static_fence();
-  static void impl_static_fence(const std::string& name);
+  static void impl_static_fence(
+      const std::string& name =
+          "Kokkos::Threads::impl_static_fence: Unnamed Static Fence");
```
Comment (dalg24) [inline core/src/Kokkos_Threads.hpp:112]: Note that would change the label that is currently printed...
https://github.com/kokkos/kokkos/blob/a793105178a9516077cdff37c07bdab2f8ba17b8/core/src/Threads/Kokkos_ThreadsExec.cpp#L298-L300

File: core/src/Kokkos_SYCL.hpp [hunk 2]
```diff
@@ -111,10 +111,12 @@ class SYCL {
   static bool wake();
 
   /** \brief Wait until all dispatched functors complete. A noop for OpenMP. */
-  static void impl_static_fence();
-  static void impl_static_fence(const std::string&);
-  void fence() const;
-  void fence(const std::string&) const;
+  static void impl_static_fence(
+      const std::string& name =
+          "Kokkos::Experimental::SYCL::fence: Unnamed Instance Fence");
```
Comment (dalg24) [inline core/src/Kokkos_SYCL.hpp:116]: Label is wrong (copied from our current implementation)

**Comment by romintomasetti** [comment]
Hi @dalg24!

We wonder with @maartenarnst why you chose to have a default value for the fence `name` argument.

As we understand it, the argument `name` is ultimately used only if the callback for the fence event in `Kokkos Tools` is activated.

Therefore, as long as the callback is `nullptr`, the argument `name` is useless.

So here is my 2 cent. We were making micro-benchmarks and were surprised of the results.

It appeared the culprit was a line `space.fence();`. In fact, we were not enabling `Kokkos Tools`, but still `space.fence();` would allocate the default `name` value (pointlessly since the callback is `nullptr`). Using `cudaStreamSynchronize(space.cuda_stream());` made a sensible difference in that particular case (in the order of the µs).

So my question is: Is there any chance that this could be refactored?

**Comment by dalg24** [comment]
I would be open to discuss if you share your benchmark.
The work you commented on was only about bringing some sanity in our internals.

**Comment by romintomasetti** [comment]
@dalg24 I made a benchmark that looks like:
```c++
bool test_fence_with_kokkos(::benchmark::State&)
{
    ...
    for( 100 times )
        space.fence(); // will use the default message, and will use Kokkos (+ expect some Kokkos Tools related overhead even when it's not used)
    ...
}

bool test_fence_backend_native(::benchmark::State&)
{
    ...
    for( 100 times)
        KOKKOS_IMPL_CUDA_SAFE_CALL(cudaStreamSynchronize(space.cuda_stream())); // backend "raw" fence, here shown for Cuda
    ...
}

bool allocate_fence_message(::benchmark::State&) const
{
    std::string test("Kokkos::Cuda::fence(): Unnamed Instance Fence");
    ::benchmark::ClobberMemory();
    return test.size() > 0;
}
```

Results, for my `AMPERE86` GPU, release mode (`g++-12`):
```bash
----------------------------------------------------------------------------------------
Benchmark                                              Time             CPU   Iterations
----------------------------------------------------------------------------------------
SpaceFenceBenchmark/fence_wrapped/manual_time      68941 ns        79466 ns         7887
SpaceFenceBenchmark/fence_raw/manual_time          53325 ns        63913 ns        13256
SpaceFenceBenchmark/allocate_fence_message          11.1 ns         11.1 ns     63001969
```

So, the overhead of using `Kokkos` to fence is somewhere around `(68941-53325)/100 ≃ 689 ns`. It seems this is explained by the allocation of the default message, but not only. I guess we can attribute the additional cost to the function calls that happen in `Kokkos::<...>::fence` (like calls to Kokkos Tools, happening *e.g.* in `Kokkos::Impl::cuda_stream_synchronize`).

Should we move this discussion somewhere else ?

**Comment by dalg24** [comment]
> Should we move this discussion somewhere else ?

Yes definitely

**Pull request closed without merging**